Doc/library/typing.rst

```diff
@@ -1266,213 +1322,320 @@ These can be used as types in annotations using ``[]``, each having a unique syn
+   using a single asterisk (``*``) before the name::
+
+      def move_first_element_to_last[T, *Ts](tup: tuple[T, *Ts]) -> tuple[*Ts, T]:
+         return (*tup[1:], tup[0])
+
+   Or by explicitly invoking the ``TypeVarTuple`` constructor::
+
+      T = TypeVar("T")
+      Ts = TypeVarTuple("Ts")
+
+      def move_first_element_to_last(tup: tuple[T, *Ts]) -> tuple[*Ts, T]:
+         return (*tup[1:], tup[0])
+
+   A normal type variable enables parameterization with a single type. A type
+   variable tuple, in contrast, allows parameterization with an
+   *arbitrary* number of types by acting like an *arbitrary* number of type
+   variables wrapped in a tuple. For example::
 
-    In contrast to non-unpacked annotations of ``*args`` - e.g. ``*args: int``,
-    which would specify that *all* arguments are ``int`` - ``*args: *Ts``
-    enables reference to the types of the *individual* arguments in ``*args``.
-    Here, this allows us to ensure the types of the ``*args`` passed
-    to ``call_soon`` match the types of the (positional) arguments of
-    ``callback``.
+      # T is bound to int, Ts is bound to ()
+      # Return value is (1,), which has type tuple[int]
+      move_first_element_to_last(tup=(1,))
 
-    See :pep:`646` for more details on type variable tuples.
+      # T is bound to int, Ts is bound to (str,)
+      # Return value is ('spam', 1), which has type tuple[str, int]
+      move_first_element_to_last(tup=(1, 'spam'))
 
-    .. versionadded:: 3.11
+      # T is bound to int, Ts is bound to (str, float)
+      # Return value is ('spam', 3.0, 1), which has type tuple[str, float, int]
+      move_first_element_to_last(tup=(1, 'spam', 3.0))
+
+      # This fails to type check (and fails at runtime)
+      # because tuple[()] is not compatible with tuple[T, *Ts]
+      # (at least one element is required)
+      move_first_element_to_last(tup=())
+
+   Note the use of the unpacking operator ``*`` in ``tuple[T, *Ts]``.
+   Conceptually, you can think of ``Ts`` as a tuple of type variables
+   ``(T1, T2, ...)``. ``tuple[T, *Ts]`` would then become
+   ``tuple[T, *(T1, T2, ...)]``, which is equivalent to
+   ``tuple[T, T1, T2, ...]``. (Note that in older versions of Python, you might
+   see this written using :data:`Unpack <Unpack>` instead, as
+   ``Unpack[Ts]``.)
+
+   Type variable tuples must *always* be unpacked. This helps distinguish type
+   variable tuples from normal type variables::
+
+      x: Ts          # Not valid
+      x: tuple[Ts]   # Not valid
+      x: tuple[*Ts]  # The correct way to do it
+
+   Type variable tuples can be used in the same contexts as normal type
+   variables. For example, in class definitions, arguments, and return types::
+
+      class Array[*Shape]:
+         def __getitem__(self, key: tuple[*Shape]) -> float: ...
+         def __abs__(self) -> "Array[*Shape]": ...
+         def get_shape(self) -> tuple[*Shape]: ...
```

Comment:
It seems like all code examples you have added or moved, the class or function bodies are only indented 3 spaces. This class has its methods indented four spaces on the `main` branch:

https://github.com/python/cpython/blob/e295d8605699ad3d8ec46c8d55a5e47da05b20c6/Doc/library/typing.rst?plain=1#L1424-L1427

Was this deliberate?

Reply:
I guess I did that because indentation for the docs themselves is by 3 spaces. We should indent by 4 in the code samples though, I'll change that.